samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -78,38 +77,38 @@
    * Allows sending {@link MessageEnvelope}s in this {@link MessageStream} to an output
    * {@link org.apache.samza.system.SystemStream} using the provided {@link SinkFunction}.
    *
-   * @param sinkFn  the function to send {@link MessageEnvelope}s in this stream to output systems
+   * @param sinkFn the function to send {@link MessageEnvelope}s in this stream to output systems
    */
   void sink(SinkFunction<M> sinkFn);
 
   /**
-   * Groups the {@link MessageEnvelope}s in this {@link MessageStream} according to the provided {@link Window} semantics
+   * Groups and processes the {@link MessageEnvelope}s in this {@link MessageStream} according to the provided {@link Window}
    * (e.g. tumbling, sliding or session windows) and returns the transformed {@link MessageStream} of
-   * {@link WindowOutput}s.
+   * {@link WindowPane}s.
    * <p>
    * Use the {@link org.apache.samza.operators.windows.Windows} helper methods to create the appropriate windows.
    *
-   * @param window  the {@link Window} to group and process {@link MessageEnvelope}s from this {@link MessageStream}
-   * @param <WK>  the type of key in the {@link WindowOutput} from the {@link Window}
-   * @param <WV>  the type of value in the {@link WindowOutput} from the {@link Window}
-   * @param <WS>  the type of window state kept in the {@link Window}
-   * @param <WM>  the type of {@link WindowOutput} in the transformed {@link MessageStream}
-   * @return  the transformed {@link MessageStream}
+   * @param window the window to group and process {@link MessageEnvelope}s from this {@link MessageStream}
+   * @param <K> the type of key in the {@link MessageEnvelope} in this {@link MessageStream}. If a key is specified,
+   *            panes are emitted per-key.
+   * @param <WK> the type of key in the {@link WindowPane} in the transformed {@link MessageStream}
+   * @param <WV> the type of value in the {@link WindowPane} in the transformed {@link MessageStream}
+   * @param <WM> the type of {@link WindowPane} in the transformed {@link MessageStream}
+   * @return the transformed {@link MessageStream}
    */
-  <WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> MessageStream<WM> window(
-      Window<M, WK, WV, WM> window);
+  <K, WK, WV, WM extends WindowPane<WK, WV>> MessageStream<WM> window(Window<M, K, WK, WV, WM> window);
```

Comment:
Yup, you're right. 

Also, On second thought, decided to use a `String` as the windowId in the `WindowKey` class.